doc/reference/reference_lua/checks.rst

```diff
@@ -0,0 +1,506 @@
+~~~~~~~~~~~~~~
+
+To make any of the supported types optional, prefix its name with a question mark (``?``).
+In the example below, the ``name`` argument is optional.
+This means that the ``greet`` function can accept ``string`` and ``nil`` values.
+
+..  literalinclude:: /code_snippets/test/checks/checks_string_optional_test.lua
+    :language: lua
+    :end-before: -- Tests
+    :dedent:
+
+As for a specific type, you can make a :ref:`union type <checks_union_type>` value optional: ``?number|string``.
+
+
+.. _checks_any_type:
+
+Turning off argument checking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can turn off checking the given arguments using the question mark (``?``) placeholder.
+In this case, the argument can be any type.
+
+..  literalinclude:: /code_snippets/test/checks/checks_any_optional_test.lua
+    :language: lua
+    :end-before: -- Tests
+    :dedent:
+
+
+
+
+.. _checks_table_type_qualifiers:
+
+Table type qualifier
+--------------------
+
+The type qualifier may be a table. In this case, the following checks are made:
+
+-   The argument is checked to conform to the ``?table`` type, and its content is validated.
+-   Table values are validated against the specified :ref:`string type qualifiers <checks_string_type_qualifier>`.
+-   Table keys missing in ``checks`` are validated against the ``nil`` type.
+
+The code below checks that the first and second table values have the ``string`` and ``number`` types.
+
+..  literalinclude:: /code_snippets/test/checks/checks_table_indexes_test.lua
+    :language: lua
+    :end-before: -- Tests
+    :dedent:
+
+In the next example, the same checks are made for the specified keys.
+
+..  literalinclude:: /code_snippets/test/checks/checks_table_keys_test.lua
+    :language: lua
+    :end-before: -- Tests
+    :dedent:
+
+.. NOTE::
+
+    A table type qualifier may be nested and use tables, too.
+
+
+
+.. _checks_api_reference:
+
+API Reference
+-------------
+
+..  container:: table
+
+    ..  rst-class:: left-align-column-1
+    ..  rst-class:: left-align-column-2
+
+    ..  list-table::
+        :widths: 35 65
+
+        *   -   **Functions**
+            -
+
+        *   -   :ref:`checks <checks-checks>`
+            -   When called inside a function, checks that the function's arguments conform to the specified types
+
+        *   -   **Objects**
+            -
+
+        *   -   :ref:`checkers <checks-checkers>`
+            -   Checkers for different types
+
+
+..  _checks-checks:
+
+checks()
+~~~~~~~~
+
+..  function:: checks(type_1, ...)
+
+    When called inside a function, checks that the function's arguments conform to the specified types.
+
+    :param string/table type_1: a :ref:`string <checks_string_type_qualifier>` or :ref:`table <checks_table_type_qualifiers>` type qualifier used to check the argument type
+    :param ...: optional type qualifiers used to check the types of :ref:`other arguments <checks_number_of_arguments>`
+
+..  _checks-checkers:
+
+checkers
```

Comment:
The only reason that `checkers` is public is that so user may add its own custom checkers. Existing checkers may be used as components for custom checkers, but there is no any other reason for one to call them explicitly. Maybe we should say that one shouldn't use `assert(checkers.datetime(arg))` instead of `checks('datetime')`

Reply:
Updated the description to clarify that its main purpose is adding a custom checker: https://docs.d.tarantool.io/en/doc/embedded-checks/reference/reference_lua/checks/#checkers